File: 51.N皇后.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<bool> col(n, false);
        // 主对角线，有行号+列号=常数
        vector<bool> left(2 * n - 1, false);
        // 次对角线，有行号-列号=常数
        vector<bool> right(2 * n - 1, false);
        vector<string> queen(n);
        string s = "";
        for(int i = 0; i < n; ++i)
            s += '.';
        for(int i = 0; i < n; ++i)
            queen[i] = s;
        dfs(col, left, right, queen, 0, n, res);
        return res;
    }

    void dfs(vector<bool> col, vector<bool> left, vector<bool> right, vector<string> queen, int row, int n, vector<vector<string>>& res){
        if(row == n) {
            res.push_back(queen);
            return;
        }
        for(int c = 0; c < n; ++c){
            if(!col[c] && !left[row + c] && !right[row - c + n - 1]){
                queen[row][c] = 'Q';
                col[c] = true;
                left[row + c] = true;
                right[row - c + n - 1] = true;
                dfs(col, left, right, queen, row + 1, n, res);
                queen[row][c] = '.';
                col[c] = false;
                left[row + c] = false;
                right[row - c + n - 1] = false;
            }
        }
    }
};

File: 51.N皇后.js
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function (n) {
    let res = [],
        col = new Array(n).fill(false), // 记录列是否有皇后
        left = new Array(2 * n - 1).fill(false), // 记录主对角线是否有皇后
        right = new Array(2 * n - 1).fill(false); // 记录次对角线是否有皇
    let queen = new Array(n).fill(0).map(() => {
        return new Array(n).fill('.')
    });
    const dfs = (row, queen) => {
        if (row === n) {
            let temp = [];
            for (let q of queen) {
                temp.push(q.join(''));
            }
            res.push(temp);
            return;
        }
        for (let c = 0; c < n; ++c) {
            // 同行列同主次对角线都没有皇后
            if (!col[c] && !left[row + c] && !right[row - c + n - 1]) {
                queen[row][c] = 'Q';
                col[c] = true;
                left[row + c] = true;
                right[row - c + n - 1] = true;
                dfs(row + 1, queen);
                right[row - c + n - 1] = false;
                left[row + c] = false;
                col[c] = false;
                queen[row][c] = '.';
            }
        }
    }

    dfs(0, queen);
    return res;
};